src/ApparentHorizons/ChangeCenterOfStrahlkorper.cpp

```diff
@@ -0,0 +1,158 @@
+    // center.
+    const std::array<double, 3> x_old{
+        {r_new * get<0>(r_hat)[s] + center_difference[0],
+         r_new * get<1>(r_hat)[s] + center_difference[1],
+         r_new * get<2>(r_hat)[s] + center_difference[2]}};
+
+    // Find (r, theta, phi) of the same point with respect to the old_center.
+    const double r_old =
+        sqrt(square(x_old[0]) + square(x_old[1]) + square(x_old[2]));
+    const double theta_old = acos(x_old[2] / r_old);
+    const double phi_old = atan2(x_old[1], x_old[0]);
+
+    // Evaluate the radius of the surface on the old strahlkorper
+    // at theta_old, phi_old.
+    const double r_surf_old = strahlkorper->radius(theta_old, phi_old);
+
+    // If r_surf_old is r_old, then 'r_new' is on the surface.
+    return r_surf_old - r_old;
+  };
+
+  const auto radius_at_each_angle = RootFinder::toms748(
+      radius_function, make_with_value<DataVector>(get<0>(r_hat), r_min * 0.9),
+      make_with_value<DataVector>(get<0>(r_hat), r_max * 1.1),
+      std::numeric_limits<double>::epsilon() * (r_min + r_max),
+      2.0 * std::numeric_limits<double>::epsilon());
+
+  // Now reset the radius and center of the new strahlkorper.
+  *strahlkorper =
+      Strahlkorper<Frame>(strahlkorper->l_max(), strahlkorper->m_max(),
+                          radius_at_each_angle, new_center);
+}
+}  // namespace change_expansion_center_of_strahlkorper_detail
+
+template <typename Frame>
+void change_expansion_center_of_strahlkorper(
+    gsl::not_null<Strahlkorper<Frame>*> strahlkorper,
+    const std::array<double, 3>& new_center) noexcept {
+  change_expansion_center_of_strahlkorper_detail::change_expansion_center(
+      strahlkorper, new_center,
+      change_expansion_center_of_strahlkorper_detail::get_rhat(*strahlkorper));
+}
+
+template <typename Frame>
+void change_expansion_center_of_strahlkorper_to_physical(
+    gsl::not_null<Strahlkorper<Frame>*> strahlkorper,
+    const double center_tolerance) noexcept {
+  const auto r_hat =
+      change_expansion_center_of_strahlkorper_detail::get_rhat(*strahlkorper);
+
+  // Zeroth iteration.
+  change_expansion_center_of_strahlkorper_detail::change_expansion_center(
+      strahlkorper, strahlkorper->physical_center(), r_hat);
+
+  // We might need more iterations until convergence.
+  const size_t maxiter = 10;
+  for (size_t iter = 0; iter < maxiter; ++iter) {
+    const auto phys_center = strahlkorper->physical_center();
+    const auto center = strahlkorper->center();
+    const double maxerr = sqrt(square(center[0] - phys_center[0]) +
+                               square(center[1] - phys_center[1]) +
+                               square(center[2] - phys_center[2]));
+    if (maxerr <= center_tolerance) {
+      return;
+    }
+    change_expansion_center_of_strahlkorper_detail::change_expansion_center(
+        strahlkorper, phys_center, r_hat);
+  }
+  ERROR("Too many iterations in change_expansion_center_of_strahlkorper");
```

Comment:
How do you know how many iterations you'll need? I worry that this could be a point of failure in our simulations...how does spec deal with this?

Reply:
SpEC sets maxiter to 100 and we never looked back.  But I don't think we need nearly that number of iterations. Convergence is fast because the approximation made by physical_center() is based on zeroing the l=1 modes, which makes smaller changes in the other modes that affect the center.